source/common/common/assert.cc

```diff
@@ -28,15 +28,63 @@ class ActionRegistrationImpl : public ActionRegistration {
   static std::function<void()> debug_assertion_failure_record_action_;
 };
 
+class EnvoyBugRegistrationImpl : public ActionRegistration {
+public:
+  EnvoyBugRegistrationImpl(std::function<void()> action) {
+    ASSERT(envoy_bug_failure_record_action_ == nullptr);
+    envoy_bug_failure_record_action_ = action;
+  }
+
+  ~EnvoyBugRegistrationImpl() override {
+    ASSERT(envoy_bug_failure_record_action_ != nullptr);
+    envoy_bug_failure_record_action_ = nullptr;
+  }
+
+  static bool shouldLogAndInvoke() {
+    ++count_;
+    if ((count_ & (count_ - 1)) == 0) {
```

Comment:
Comment since the implementation of this predicate is a bit subtle.

Reply:
I'm assuming we won't log over 2**(2**64) ENVOY_BUGs, but we we do we miss the 0 case after wrap-around.